From dc3e3547bae9181ddf2698e97b88b5ac2f3d225f Mon Sep 17 00:00:00 2001
From: Manish Dharanenthiran <quic_mdharane@quicinc.com>
Date: Wed, 14 Dec 2022 12:01:35 +0530
Subject: [PATCH] ath12k: Update pending desc size in reo cache flush

During rx_tid_del, if the desc size is greater than
128 bytes, total desc size will be splitted into
respetive 128 byte low and high address format and
updated in flush cache descriptor. During failure
case, we will have the entire rx_tid will be saved
to reo_cmd list to free in next cycle. In a scenario,
where one 128 byte desc is actually freed and the
next desc is failed, we are still saving the entrie
rx_tid and in the next cycle we will try to flush
the same desc again which results in memory corruption.

To fix this, update pending desc size to rx_tid struct
and use that in next cycle while calculating total
desc size in order to avoid updating same desc again.

Added two soc_dp_stats variable to capture reo_cmd
send failure and drain cmd stats.

Signed-off-by: Manish Dharanenthiran <quic_mdharane@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/core.h    |  2 ++
 drivers/net/wireless/ath/ath12k/debugfs.c |  8 ++++++++
 drivers/net/wireless/ath/ath12k/dp_rx.c   | 19 +++++++++----------
 drivers/net/wireless/ath/ath12k/dp_rx.h   |  1 +
 4 files changed, 20 insertions(+), 10 deletions(-)

diff --git a/drivers/net/wireless/ath/ath12k/core.h b/drivers/net/wireless/ath/ath12k/core.h
index 6d12cc5..4a9702f 100644
--- a/drivers/net/wireless/ath/ath12k/core.h
+++ b/drivers/net/wireless/ath/ath12k/core.h
@@ -981,6 +981,8 @@ struct ath12k_soc_dp_stats {
 	u32 reo_error_drop[HAL_REO_DEST_RING_ERROR_CODE_MAX];
 	u32 hal_reo_error[DP_REO_DST_RING_MAX];
 	u32 mon_drop_desc;
+	u32 hal_reo_cmd_drain;
+	u32 reo_cmd_cache_error;
 	struct ath12k_soc_dp_tx_err_stats tx_err;
 	struct ath12k_dp_ring_bp_stats bp_stats;
 };
diff --git a/drivers/net/wireless/ath/ath12k/debugfs.c b/drivers/net/wireless/ath/ath12k/debugfs.c
index 55924ec..f121252 100644
--- a/drivers/net/wireless/ath/ath12k/debugfs.c
+++ b/drivers/net/wireless/ath/ath12k/debugfs.c
@@ -810,6 +810,14 @@ static ssize_t ath12k_debugfs_dump_soc_dp_stats(struct file *file,
 			 "\nMon drop descriptor: %u\n",
 			 soc_stats->mon_drop_desc);
 
+	len += scnprintf(buf + len, size - len,
+			"\nHAL_REO_CMD_DRAIN Counter: %u\n",
+			soc_stats->hal_reo_cmd_drain);
+
+	len += scnprintf(buf + len, size - len,
+			"\nREO_CMD_CACHE_FLUSH Failure: %u\n",
+			soc_stats->reo_cmd_cache_error);
+
 	len += ath12k_debugfs_dump_soc_ring_bp_stats(ab, buf + len, size - len);
 
 	if (len > size)
diff --git a/drivers/net/wireless/ath/ath12k/dp_rx.c b/drivers/net/wireless/ath/ath12k/dp_rx.c
index 31462e7..0a816ff 100644
--- a/drivers/net/wireless/ath/ath12k/dp_rx.c
+++ b/drivers/net/wireless/ath/ath12k/dp_rx.c
@@ -699,7 +699,10 @@ static int ath12k_dp_reo_cache_flush(struct ath12k_base *ab,
 	unsigned long tot_desc_sz, desc_sz;
 	int ret;
 
-	tot_desc_sz = rx_tid->size;
+	if (rx_tid->pending_desc_size)
+		tot_desc_sz = rx_tid->pending_desc_size;
+	else
+		tot_desc_sz = rx_tid->size;
 	desc_sz = ath12k_hal_reo_qdesc_size(0, HAL_DESC_REO_NON_QOS_TID);
 
 	while (tot_desc_sz > desc_sz) {
@@ -711,9 +714,7 @@ static int ath12k_dp_reo_cache_flush(struct ath12k_base *ab,
 					     NULL);
 		if (ret)
 		{
-			ath12k_warn(ab,
-				    "failed to send HAL_REO_CMD_FLUSH_CACHE, tid %d (%d) desc_sz(%ld)\n",
-				    rx_tid->tid, ret, desc_sz);
+			rx_tid->pending_desc_size = tot_desc_sz + desc_sz;
 			/* If this fails with ring full condition, then
 			 * no need to retry below as it is expected to
 			 * fail within short time */
@@ -722,6 +723,7 @@ static int ath12k_dp_reo_cache_flush(struct ath12k_base *ab,
 		}
 	}
 
+	rx_tid->pending_desc_size = desc_sz;
 	memset(&cmd, 0, sizeof(cmd));
 	cmd.addr_lo = lower_32_bits(rx_tid->paddr);
 	cmd.addr_hi = upper_32_bits(rx_tid->paddr);
@@ -729,9 +731,6 @@ static int ath12k_dp_reo_cache_flush(struct ath12k_base *ab,
 	ret = ath12k_dp_reo_cmd_send(ab, rx_tid,
 				     HAL_REO_CMD_FLUSH_CACHE,
 				     &cmd, ath12k_dp_reo_cmd_free);
-	if (ret)
-		ath12k_err(ab, "failed to send HAL_REO_CMD_FLUSH_CACHE cmd, tid %d (%d)\n",
-			   rx_tid->tid, ret);
 
 exit:
 	return ret;
@@ -745,6 +744,7 @@ static void ath12k_dp_rx_tid_del_func(struct ath12k_dp *dp, void *ctx,
 	struct ath12k_dp_rx_reo_cache_flush_elem *elem, *tmp;
 
 	if (status == HAL_REO_CMD_DRAIN) {
+		ab->soc_stats.hal_reo_cmd_drain++;
 		goto free_desc;
 	} else if (status != HAL_REO_CMD_SUCCESS) {
 		/* Shouldn't happen! Cleanup in case of other failure? */
@@ -777,12 +777,11 @@ static void ath12k_dp_rx_tid_del_func(struct ath12k_dp *dp, void *ctx,
 			if (ath12k_dp_reo_cache_flush(ab, &elem->data)) {
 				/* In failure case, just update the timestamp
 				 * for flush cache elem and continue */
+				ab->soc_stats.reo_cmd_cache_error++;
 				spin_lock_bh(&dp->reo_cmd_lock);
 				elem->ts = jiffies +
 					msecs_to_jiffies(ATH12K_DP_RX_REO_DESC_FREE_TIMEOUT_MS);
-				ath12k_warn(ab, "Failed to send HAL_REO_CMD_FLUSH_CACHE cmd"
-						"Updating timestamp (%ld) in the list\n", elem->ts);
-				continue;
+				break;
 			}
 			spin_lock_bh(&dp->reo_cmd_lock);
 			list_del(&elem->list);
diff --git a/drivers/net/wireless/ath/ath12k/dp_rx.h b/drivers/net/wireless/ath/ath12k/dp_rx.h
index e6095b8..9cecfff 100644
--- a/drivers/net/wireless/ath/ath12k/dp_rx.h
+++ b/drivers/net/wireless/ath/ath12k/dp_rx.h
@@ -17,6 +17,7 @@ struct ath12k_dp_rx_tid {
 	u32 *vaddr;
 	dma_addr_t paddr;
 	u32 size;
+	u32 pending_desc_size;
 	u32 ba_win_sz;
 	bool active;
 
-- 
2.17.1

